File: backend/src/services/monitoring.js
import { createClient } from 'redis';
import winston from 'winston';

const logger = winston.createLogger({
  transports: [new winston.transports.Console()]
});

const REDIS_URL = process.env.REDIS_URL || "redis://localhost:6379/0";
let redisClient = null;

if (process.env.ENVIRONMENT === 'production' || process.env.REDIS_URL) {
  try {
    redisClient = createClient({ url: REDIS_URL });
    redisClient.on('error', () => {});
    await redisClient.connect().catch(() => {
      redisClient = null;
    });
  } catch (err) {
    redisClient = null;
  }
}

// In-memory counters fallback
const memoryMetrics = new Map();

/**
 * Increment custom security metric counter (Layer 17.1)
 */
export async function incrementSecurityMetric(name, labels = {}) {
  const labelParts = Object.entries(labels)
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(([k, v]) => `${k}=${v}`);
  const labelStr = labelParts.length > 0 ? `:${labelParts.join(',')}` : "";
  const redisKey = `metrics:${name}${labelStr}`;

  if (redisClient) {
    try {
      await redisClient.incr(redisKey);
      return;
    } catch (e) {
      logger.error(`Redis metrics increment failed: ${e.message}`);
    }
  }

  // Memory fallback
  const count = (memoryMetrics.get(redisKey) || 0) + 1;
  memoryMetrics.set(redisKey, count);
}

/**
 * Compile all metrics in Prometheus format (Layer 17.1)
 */
export async function getMetricsPrometheusFormat() {
  const lines = [];

  try {
    if (redisClient) {
      const keys = await redisClient.keys("metrics:*");
      for (const key of keys) {
        const value = await redisClient.get(key);
        const parts = key.split(":");
        const metricName = parts[1];
        
        let labelStr = "";
        if (parts.length > 2 && parts[2]) {
          const labels = {};
          parts[2].split(",").forEach(pair => {
            if (pair.includes("=")) {
              const [k, v] = pair.split("=");
              labels[k] = v;
            }
          });
          labelStr = "{" + Object.entries(labels).map(([k, v]) => `${k}="${v}"`).join(",") + "}";
        }

        lines.push(`# TYPE ${metricName} counter`);
        lines.push(`${metricName}${labelStr} ${value}`);
      }
    } else {
      for (const [key, value] of memoryMetrics.entries()) {
        const parts = key.split(":");
        const metricName = parts[1];
        
        let labelStr = "";
        if (parts.length > 2 && parts[2]) {
          const labels = {};
          parts[2].split(",").forEach(pair => {
            if (pair.includes("=")) {
              const [k, v] = pair.split("=");
              labels[k] = v;
            }
          });
          labelStr = "{" + Object.entries(labels).map(([k, v]) => `${k}="${v}"`).join(",") + "}";
        }

        lines.push(`# TYPE ${metricName} counter`);
        lines.push(`${metricName}${labelStr} ${value}`);
      }
    }
  } catch (err) {
    logger.error(`Failed to compile prometheus metrics: ${err.message}`);
  }

  return lines.join("\n");
}

/**
 * Fetch aggregated security metrics for health reporting (Layer 17.2)
 */
export async function getSecurityHealthStatus() {
  let bruteForceBlocks = 0;
  let suspiciousIps = 0;
  let rateLimits = 0;

  try {
    const scanKeys = async (pattern) => {
      let sum = 0;
      if (redisClient) {
        const keys = await redisClient.keys(`metrics:${pattern}*`);
        for (const k of keys) {
          const val = await redisClient.get(k);
          sum += parseInt(val || '0', 10);
        }
      } else {
        for (const [k, val] of memoryMetrics.entries()) {
          if (k.startsWith(`metrics:${pattern}`)) {
            sum += val;
          }
        }
      }
      return sum;
    };

    bruteForceBlocks = await scanKeys("flowshield_account_lockouts_total");
    suspiciousIps = await scanKeys("flowshield_suspicious_ips_blocked_total");
    rateLimits = await scanKeys("flowshield_rate_limit_hits_total");
  } catch (err) {
    // Ignore and fallback to 0
  }

  return {
    brute_force_blocks_today: bruteForceBlocks,
    suspicious_ips_blocked: suspiciousIps,
    rate_limit_violations_today: rateLimits,
    last_security_scan: new Date().toISOString()
  };
}

File: backend/app/core/monitoring.py
import os
import logging
from typing import Dict, Optional
import redis.asyncio as async_redis

logger = logging.getLogger(__name__)

REDIS_URL = os.getenv("REDIS_URL", "redis://localhost:6379/0")
async_redis_client = async_redis.from_url(REDIS_URL, decode_responses=True)

async def increment_security_metric(name: str, labels: Optional[Dict[str, str]] = None):
    """Increment custom security metric counter in Redis."""
    try:
        label_parts = []
        if labels:
            for k, v in sorted(labels.items()):
                label_parts.append(f"{k}={v}")
        label_str = f":{','.join(label_parts)}" if label_parts else ""
        
        # Redis key format: metrics:metric_name:label_name=label_value
        redis_key = f"metrics:{name}{label_str}"
        await async_redis_client.incr(redis_key)
    except Exception as e:
        logger.error(f"Failed to increment metric {name}: {e}")

async def get_metrics_prometheus_format() -> str:
    """Return all custom metrics in standard Prometheus exporter format."""
    lines = []
    try:
        async for key in async_redis_client.scan_iter("metrics:*"):
            value = await async_redis_client.get(key)
            # parse key
            parts = key.split(":")
            metric_name = parts[1]
            labels = {}
            if len(parts) > 2 and parts[2]:
                for pair in parts[2].split(","):
                    if "=" in pair:
                        k, v = pair.split("=", 1)
                        labels[k] = v
                        
            label_str = ""
            if labels:
                label_str = "{" + ",".join(f'{k}="{v}"' for k, v in labels.items()) + "}"
            
            lines.append(f"# TYPE {metric_name} counter")
            lines.append(f"{metric_name}{label_str} {value}")
    except Exception as e:
        logger.error(f"Failed to compile prometheus metrics: {e}")
    return "\n".join(lines)

async def get_security_health_status() -> dict:
    """Fetch aggregated security metrics for health status report."""
    # Read metrics counts from Redis
    brute_force_blocks = 0
    suspicious_ips = 0
    rate_limits = 0
    
    try:
        # Scan keys to sum them
        async for key in async_redis_client.scan_iter("metrics:flowshield_account_lockouts_total*"):
            val = await async_redis_client.get(key)
            brute_force_blocks += int(val or 0)
            
        async for key in async_redis_client.scan_iter("metrics:flowshield_suspicious_ips_blocked_total*"):
            val = await async_redis_client.get(key)
            suspicious_ips += int(val or 0)

        async for key in async_redis_client.scan_iter("metrics:flowshield_rate_limit_hits_total*"):
            val = await async_redis_client.get(key)
            rate_limits += int(val or 0)
    except Exception:
        pass
        
    return {
        "brute_force_blocks_today": brute_force_blocks,
        "suspicious_ips_blocked": suspicious_ips,
        "rate_limit_violations_today": rate_limits,
        "last_security_scan": "2026-06-22T09:00:00Z"
    }
